src/TypeShape/Applications/UnionContract.fs

```diff
@@ -0,0 +1,206 @@
+﻿module TypeShape.UnionContract
+
+open System
+open System.Runtime.Serialization
+open TypeShape.Core
+open TypeShape.Core.Utils
+
+/// Marker interface for tagging union contract types
+type IUnionContract = interface end
+
+/// Generic encoding abstraction for serializing/deserializing 
+/// to a fixed format type, e.g. string, byte[], Newtonsoft JsonValue etc
+type IEncoder<'Format> =
+    /// 'Null' format value to be used for union cases without a payload
+    abstract Empty : 'Format
+    /// Encodes any value to the implementation format
+    abstract Encode<'T> : value:'T -> 'Format
+    /// Decodes format instance to any value
+    abstract Decode<'T> : fmt:'Format -> 'T
+
+/// Generic encoder that simply upcasts values to System.Object
+type CastEncoder() =
+    interface IEncoder<obj> with
+        member __.Empty    = null
+        member __.Encode t = box t
+        member __.Decode o = unbox o
+
+/// Represents an encoded union case
+type EncodedUnion<'Encoding> =
+    {
+        CaseName : string
+        Payload  : 'Encoding
+    }
+
+/// Provides an encoder implementation for a sum of events
+[<AbstractClass>]
+type UnionContractEncoder<'Union, 'Format when 'Union :> IUnionContract> internal () =
+    /// Gets the union case string identifier for given union instance
+    abstract GetCaseName : value:'Union -> string
+    /// Encodes a union instance into a decoded representation
+    abstract Encode      : value:'Union -> EncodedUnion<'Format>
+    /// Decodes a formatted representation into a union instance. Can potentially throw FormatException
+    abstract Decode      : encodedUnion:EncodedUnion<'Format> -> 'Union
+    /// Decodes a formatted representation into a union instance. Does not throw exception on format mismatches
+    abstract TryDecode   : encodedUnion:EncodedUnion<'Format> -> 'Union option
+
+[<AutoOpen>]
+module private Impl =
+
+    type Config = { requireRecordPayloads : bool ; allowNullaryCases : bool }
```

Comment:
can you put xmldoc on these ? (reason I ask is that I'm asking myself as I read: should those be mutually exclusive and/or a DU?)

Reply:
These are private implementation types

Reply:
doesn't mean it can't have a comment!